Feature/pass args to faker

mastacheata opened the pull request:
Hello,

I'd like to suggest modifying the Faker scrubber to pass along any arguments to the Faker method.
Most Faker methods can be called without any parameters, but the date fakers for example allow to specify parameters to further limit the range in which dates are generated. That would be super helpful when used with Django-scrubber as well.
Also you could use this to create new fakes without havng to write a new Fake Provider directly from the django code by using the Methods from the BaseProvider (strangely documented in the Address provider).

I also changed the docstrings to use double quotes as that's what PEP257 suggests (and because my IDE complains about it :wink:):
> For consistency, always use """triple double quotes""" around docstrings.

costela commented:
Thanks for the contribution! Would you mind also expanding the `README` to include examples of this usage? And a test case would also be nice! :wink:
As for build issue, it looks like a hiccup. Restarting now.

mastacheata commented:
Sure thing.
Will add that tomorrow. Thanks for the feedback and restarting the test.

costela commented on tests/test_scrubbers.py:
```diff
@@ -29,3 +30,16 @@ def test_hash_scrubber_textfield(self):
 
         self.assertNotEqual(data.description, 'Foo')
+
+    def test_faker_scrubber_date_in_past(self):
```
nitpicking: you are actually not interested in the `date_in_past` part. You're just using that one specific provider as an example, but the test is actually about passing arbitrary arguments to faker providers.
```suggestion
    def test_faker_scrubber_with_provider_arguments(self):
```

costela commented on tests/test_scrubbers.py:
```diff
@@ -29,3 +30,16 @@ def test_hash_scrubber_textfield(self):
+        data.refresh_from_db()
+
+        self.assertGreater(datetime.date.today(), data.date_past)
+        self.assertGreater(datetime.date.today() - datetime.timedelta(days=28), data.date_past)
```
I'm not sure this is a good idea. This test might pass by chance, if the first generated fake date just _happens_ to be further in the past than 28 days. Maybe we should use some other provider for this test? One whose behavior with and without arguments is ensured to have no overlap?

Or am I missing something?

costela replied:
I mean the complementary case: the date _with parameters_ is guaranteed to be 29 days in the past, but the date _without parameters_ is *not* guaranteed to *not* be more than 29 days in the past. So the test might *pass*, even if the provider did *not* receive the parameters.

mastacheata replied:
Ahh, damn, you're right. I didn't think to test the opposite as well.
How about the ean faker.
It returns 13 digits by default and can be changed to the 8 digit version using a parameter?

mastacheata replied:
That would also not run into trouble with Postgres not casting strings to dates automatically. (which is a different issue altogether)

costela replied:
Sounds good!
Let's split this: please refactor the test then we can try a second PR to add arbitrary field types.

costela commented:
The postgresql test failures highlight one current limitation of `django-scrubber`: the model used to store the faker data has a single _text_ field. If you want to scrub different field types, you have to do explicit casting on the ORM/SQL level.
My first idea is to look at `scrubbers.Faker.__call__` and try to automatically re-cast to the destination column type. But maybe there are more elegant solutions.

mastacheata commented:
I found another problem with passing parameters:
Faker('ean', 8) != Faker('ean')
But your library will reuse an already initialized provider solely based on it's name since it's not aware of the parameters at all.

GitRon commented:
@costela Any update on the topic?

mastacheata commented:
> 
> 
> @costela Any update on the topic?

@GitRon has resolved the conflicts. (We're working for the same company on different projects. In my project we got around the parameter limitation using a different provider instead, but I guess his project does need the parameters now)

costela commented:
Hi @mastacheata, @GitRon,

Big thanks for picking this PR up again!

Unfortunately the issue you mentioned in https://github.com/RegioHelden/django-scrubber/pull/11#issuecomment-433085706 is still open. Anybody want to give it a try?

mastacheata commented:
That one is the easy part.
It's not about storing the params in the database, but about the initialized_providers set.

The bigger problem is casting the values from string to whatever field type is needed as found in the comment about the failing test for postgres or the new issue created by Ronny.

But maybe that's something he'd like to have a look at (@GitRon ?)

costela commented on django_scrubber/scrubbers.py:
```diff
@@ -100,8 +100,12 @@ def __call__(self, field):
+    def __init__(self, provider, *args, **kwargs):
         self.provider = provider
+        self.provider_args = args
+        self.provider_kwargs = kwargs
```
we probably don't need to store these if they're only used for generating the hash key?

mastacheata replied:
We need them, they're to be used in the FakeData generation (see next comment). Otherwise you can't pass the args/kwargs to the faker provider.

costela replied:
ah yes, you're right.

costela commented on django_scrubber/scrubbers.py:
```diff
@@ -136,23 +140,24 @@ def _initialize_data(self):
         faker_instance.seed(settings_with_fallback('SCRUBBER_RANDOM_SEED'))
         for i in range(settings_with_fallback('SCRUBBER_ENTRIES_PER_PROVIDER')):
             fakedata.append(FakeData(provider=self.provider, provider_offset=i,
```
you probably want `self.provider_key` here, right?

costela commented on tests/test_scrubbers.py:
```diff
@@ -38,3 +37,23 @@ def test_faker_scrubber_charfield(self):
+        data.refresh_from_db()
+
+        # make sure it doesn't reuse the ean with length=8 scrubber
+        self.assertEquals(9, len(data.ean9))
```
you probably mean `ean8` in these, right?

mastacheata replied:
> you probably mean `ean8` in these, right?

You're right. I mixed that up while replacing.
We can use both scrubbers on the same field and see if that changes the stuff, but I'm not sure if that works. Will try running that test again and fix the field names.

mastacheata replied:
Ok, it still works when using the same field for both, so I ditched the other one.

costela commented on tests/models.py:
```diff
@@ -7,6 +7,8 @@ class DataToBeScrubbed(models.Model):
     last_name = models.CharField(max_length=255)
     description = models.TextField()
+    ean8 = models.CharField(max_length=13)
+    ean13 = models.CharField(max_length=13)
```
this field isn't used?

mastacheata replied:
I'll go over that test again. Seems like I messed that up wanting to do just a quick and dirty fix.

mastacheata replied:
Removed the surplus field

costela commented:
@mastacheata 
> I've got no clue how to even find the datatype of the target field, let alone how to cast strings to arbitrary types in python.

I'm thinking maybe we can optimistically use [`Cast()`](https://docs.djangoproject.com/en/2.1/ref/models/database-functions/#cast) inside of `Faker.__call__` (since we have access to the field, and can derive its type).  
The text value returned by the faker provider would have to be something that `Cast()` can handle, and there would still be potential for mismatches, but I think this might work for many cases.

mastacheata commented:
I've implemented the optimistic casting as you suggested and fixed my tests.
At least for DateFields it seems to work just fine. (That's the TestCase I removed last year in favor of the ean faker and now readded)

Should we maybe add some more Tests for different field types?

mastacheata commented:
Hmm, looks like SQLite has a problem with the DateField.   
Apparently the DateField is not being filled with either the scrubber or the default value.

No idea why that happens, though. 🤷‍♂

costela commented:
> Apparently the DateField is not being filled with either the scrubber or the default value.

Can you please check what the contents of the `FakerData` model are? I just want to rule out that the problem is not happening while initializing the fake data.

(sorry, I don't have access to a device to test this myself right now)

GitRon commented:
@mastacheata @costela There is a bug with django < 2.1 and sqlite. As a quick fix we could add the limitation in the Readme and just run the test for not-sqlite or django > 2.1

Opinions?

GitRon commented:
@costela @mastacheata is ok with my solution. I updated the limitations section and fixed the test. If you give your ok as well, I'll create a PR to @mastacheata 's branch and then we maybe can create a new version at pypi quickly? 😃

costela commented:
@GitRon sounds OK! :+1:

mastacheata commented:
I tried to disallow null on the DateField and instead set a default value so the other tests could pass without affecting the Date test, but to no avail. That's why I think skipping that test for older Django versions is the right way. (Would waste too much time finding out why it breaks there otherwise, if someone knows about a known-problem with datefields in older django versions, they can always come and fix it)

GitRon commented:
Agreed.

GitRon commented:
@costela Pipeline runs. Do you think you could create a new release soon™? 😃

costela commented on .gitignore:
```diff
@@ -45,3 +45,6 @@ output/*/index.html
 docs/_build
+
+# Sqlite databases
+*.sqlite3
```
this is probably also a left-over from sqlite testing?

GitRon replied:
✔️

costela commented on tests/settings.py:
```diff
@@ -2,16 +2,18 @@
 DATABASES = {
-    "default": dj_database_url.config(default='sqlite://:memory:')
+    'default': {
+        'ENGINE': 'django.db.backends.sqlite3',
```
this is probably also a left-over from sqlite testing, right? This basically disables testing for mysql and postgresql.

GitRon replied:
I want to remove the dependency to `dj_database_url` which is unnecessary. Thats why I wrote it this way. Ok for you?

costela replied:
it's ok in principle, but you'd have to find a comparably easy way of [loading these settings from the environment](https://github.com/RegioHelden/django-scrubber/blob/master/tox.ini#L10-L12).
I'd suggest we leave it to a different PR and focus on the feature here.

GitRon replied:
But it works now and we need one package less. Whats the drawback? 🤔

costela replied:
It "works" because all tests are using sqlite everywhere, even those that should be using mysql or postgres. We're not testing what we think we're testing.

GitRon replied:
And with your line, it is different?

mastacheata replied:
Yeah, the tox.ini defines the database urls for the dj_database_url plugin as environment variables and apparently the plugin has predefined environment variables where it'll look for the config without explicitly reading the environment variables into the settings.py

Also I'm with @costela on this: Make single-purpose PRs and just create a new one if you'd like to suggest a different style of things. It's just a stylistic choice and there's room for arguing about that on a subjective basis, while improving the faker thing is an objective improvement which everyone seems to have agreed on already.

costela commented on tests/test_scrubbers.py:
```diff
@@ -38,3 +41,38 @@ def test_faker_scrubber_charfield(self):
+    def test_faker_scrubber_datefield(self):
+        """
+        Use this as an example for Scrubber's capability of optimistically Casting to the current field's type
+        There is a bug with django < 2.2 and sqlite, that's why we don't run the test there.
```
we're currently not testing django 2.2. Could you please also add that to the PR? Should be just a matter of adding travis and tox configs.

GitRon replied:
Thanks for pointing that out. Acutally it works in django 2.1 as well. Fixed it AND added the django 2.2 test.

GitRon commented on README.md:
```diff
@@ -123,6 +125,8 @@ This behaviour can be changed by setting `SCRUBBER_RANDOM_SEED=None`, which ensu
 Scrubbing unique fields may lead to `IntegrityError`s, since there is no guarantee that the random content will not be repeated. Playing with different settings for `SCRUBBER_RANDOM_SEED` and `SCRUBBER_ENTRIES_PER_PROVIDER` may alleviate the problem.
 Unfortunately, for performance reasons, the source data for scrubbing with faker is added to the database, and arbitrarily increasing `SCRUBBER_ENTRIES_PER_PROVIDER` will significantly slow down scrubbing (besides still not guaranteeing uniqueness).
 
+When using `django < 2.2` and working on `sqlite` a bug within django causes field-specific scrubbing (e.g. `date_object`, `boolean`) to fail. Please consider using a different database backend or upgrade to the latest django version.
```
removed boolean

GitRon commented on tests/settings.py:
```diff
@@ -2,16 +2,18 @@
 
 # SECURITY WARNING: keep the secret key used in production secret!
-SECRET_KEY = "hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh"
+SECRET_KEY = "uzbLoOIYlJnzGDYlUfynNyocjZH9NLSc3AAREwLDaugQkCzsQn"
```
because it looks more professional? 😉

costela replied:
:man_shrugging:

GitRon replied:
It's an improvement 😇

GitRon commented on tests/test_scrubbers.py:
```diff
@@ -1,11 +1,14 @@
 from django_scrubber import scrubbers
-
-from .models import DataFactory, DataToBeScrubbed
+from tests.models import DataFactory, DataToBeScrubbed
```
✔️

GitRon commented:
@costela Will give a PR to @mastacheata now.

GitRon commented:
@costela All good now? 😃

mastacheata commented on tests/test_scrubbers.py:
```diff
@@ -38,3 +41,38 @@ def test_faker_scrubber_charfield(self):
+        Use this as an example for Scrubber's capability of optimistically Casting to the current field's type
+        There is a bug with django < 2.2 and sqlite, that's why we don't run the test there.
+        """
+        if django.VERSION > (2, 1) or connection.vendor != "sqlite":
```
@GitRon if it works in 2.1, why skip the test in 2.1?

GitRon replied:
fixed

mastacheata commented on tox.ini:
```diff
@@ -16,6 +16,7 @@ deps =
     django111: Django >=1.11, <1.12
     django20: Django >=2.0, <2.1
     django21: Django >=2.1, <2.2
+    django22: Django >=2.2, <2.3
```
No idea how tox.ini works, but the latest travis run didn't include a django22 environment.

GitRon replied:
👍

costela commented:
Just did a last round of rebasing and a quick fix to missing travis settings.

@mastacheata @GitRon thanks both of you for all the work! This is a nice feature!

GitRon commented:
@costela Thanks for approving the PR! Please give us a short update when the release is out, will you?

costela commented:
build is running!
